pkg/engine/executor/dataobjscan_predicate.go

```diff
@@ -43,5 +47,221 @@ func buildLogsPredicate(_ physical.Expression) (dataobj.LogsPredicate, error) {
 	// think we should start removing things from internal for this; we can probably
 	// find a way to remove the explicit dependency from the dataobj package from
 	// the physical planner instead.
-	return nil, fmt.Errorf("logs predicate conversion is not supported")
+	return mapInitiallySupportedPredicates(expr)
+}
+
+// Support for timestamp and metadata predicates has been implemented.
+// TODO(owen-d): this can go away when we use dataset.Reader & dataset.Predicate directly
+func mapInitiallySupportedPredicates(expr physical.Expression) (dataobj.LogsPredicate, error) {
+	switch e := expr.(type) {
+	case *physical.BinaryExpr:
+		if e.Left.Type() != physical.ExprTypeColumn {
+			return nil, fmt.Errorf("unsupported predicate, expected column ref on LHS: %s", expr.String())
+		}
+
+		left := e.Left.(*physical.ColumnExpr)
+		switch left.Ref.Type {
+		case types.ColumnTypeBuiltin:
+			if left.Ref.Column == types.ColumnNameBuiltinTimestamp {
+				return mapTimestampPredicate(e)
+			}
+			return nil, fmt.Errorf("unsupported builtin column in predicate (only timestamp is supported for now): %s", left.Ref.Column)
+		case types.ColumnTypeMetadata:
+			return mapMetadataPredicate(e)
+		default:
+			return nil, fmt.Errorf("unsupported column ref type (%T) in predicate: %s", left.Ref, left.Ref.String())
+		}
+	default:
+		return nil, fmt.Errorf("unsupported expression type (%T) in predicate: %s", expr, expr.String())
+	}
+}
+
+func mapTimestampPredicate(expr physical.Expression) (dataobj.TimeRangePredicate[dataobj.LogsPredicate], error) {
+	m := newTimestampPredicateMapper()
+	if err := m.verify(expr); err != nil {
+		return dataobj.TimeRangePredicate[dataobj.LogsPredicate]{}, err
+	}
+
+	binop := expr.(*physical.BinaryExpr)
+	err := m.update(binop.Op, binop.Right)
+
+	return m.res, err
+}
+
+func newTimestampPredicateMapper() *timestampPredicateMapper {
+	open := dataobj.TimeRangePredicate[dataobj.LogsPredicate]{
+		StartTime:    time.Unix(0, math.MinInt64).UTC(),
+		EndTime:      time.Unix(0, math.MaxInt64).UTC(),
+		IncludeStart: true,
+		IncludeEnd:   true,
+	}
+	return &timestampPredicateMapper{
+		res: open,
+	}
+}
+
+type timestampPredicateMapper struct {
+	res dataobj.TimeRangePredicate[dataobj.LogsPredicate]
+}
+
+// ensures the LHS is a timestamp column reference
+// and the RHS is either a literal or a binary expression
```

Comment:
In which case is RHS a binary expression? I thought it would always be a timestamp literal. At least that's what we create in the logical planner:

https://github.com/grafana/loki/blob/main/pkg/engine/planner/logical/planner.go#L239-L252

Reply:
To simplify the conversion, we should imo either:
a. Align the `TimeRangePredicate` with physical binary expression in a way that it follows the semantics of a binary expression with an `AND` operator and a `GT(E)` operator on the LHS and a `LT(E)` operatior on the RHS.
b. Create a physcial binary expression with a `BETWEEN` operator that matches more the semantics of the `TimeRangePredicate`.